src/refiners/foundationals/latent_diffusion/ella_adapter.py

```diff
@@ -0,0 +1,294 @@
+
+
+class Latents(fl.Chain):
+    def __init__(
+        self,
+        num_latents: int,
+        width: int,
+        device: Device | str | None = None,
+        dtype: DType | None = None,
+    ) -> None:
+        super().__init__(
+            ParameterInitialized(
+                num_latents,
+                width,
+                device=device,
+                dtype=dtype,
+            ),
+        )
+
+
+class LatentsAdaLayerNorm(AdaLayerNorm):
+    pass
+
+
+class InputAdaLayerNorm(AdaLayerNorm):
+    pass
+
+
+class OutputAdaLayerNorm(AdaLayerNorm):
+    pass
+
+
+class _PerceiverAttentionBlock(fl.Chain):
+    def __init__(
+        self,
+        width: int,
+        num_heads: int,
+        timestep_embedding_dim: int,
+        device: Device | str | None = None,
+        dtype: DType | None = None,
+    ) -> None:
+        super().__init__(
+            fl.Sum(
+                fl.Parallel(
+                    fl.GetArg(0),
+                    fl.Chain(
+                        fl.Distribute(
+                            LatentsAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
+                            InputAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
+                        ),
+                        fl.Parallel(
+                            fl.GetArg(0),
+                            fl.Concatenate(fl.GetArg(0), fl.GetArg(1), dim=1),
+                            fl.Concatenate(fl.GetArg(0), fl.GetArg(1), dim=1),
+                        ),
+                        fl.Attention(embedding_dim=width, num_heads=num_heads, device=device, dtype=dtype),
+                    ),
+                )
+            ),
+            fl.Residual(
+                fl.Chain(
+                    OutputAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
+                    fl.Linear(width, width * 4, device=device, dtype=dtype),
+                    SquaredReLU(),
+                    fl.Linear(width * 4, width, device=device, dtype=dtype),
+                )
+            ),
+        )
+
+
+class PerceiverAttentionBlock(fl.Chain):
+    def __init__(
+        self,
+        width: int,
+        num_heads: int,
+        timestep_embedding_dim: int,
+        device: Device | str | None = None,
+        dtype: DType | None = None,
+    ) -> None:
+        super().__init__(
+            fl.Parallel(
+                _PerceiverAttentionBlock(width, num_heads, timestep_embedding_dim, device=device, dtype=dtype),
+                fl.GetArg(1),
+            )
+        )
+
+
+class OutputProjection(fl.Chain):
+    def __init__(
+        self, width: int, output_dim: int, device: Device | str | None = None, dtype: DType | None = None
+    ) -> None:
+        super().__init__(
+            fl.Linear(width, output_dim, device=device, dtype=dtype),
+            fl.LayerNorm(output_dim, device=device, dtype=dtype),
+        )
+
+
+class PerceiverResampler(fl.Chain):
```

Comment:
Out of curiosity: did you have look at the Perceiver Resampler already implemented part of IP-Adapter Plus? see https://github.com/finegrain-ai/refiners/blob/10fba3467ad3cd9e878ad3b7e797b2804cbbb261/src/refiners/foundationals/latent_diffusion/image_prompt.py#L182

Are there noticeable differences? If yes, are there ways to keep this new version consistent with the other one?

Reply:
I did check this implementation. except for `PerceiverScaledDotProductAttention`  They are identical actually. I tried to follow the same modulation from the official ELLA repo. But I can reformulate my perceiverResampler implementation to match the one for IP-adapter.

Reply:
> They are identical actually with few differences

OK, good to know.

In other similar cases (you might have noticed that there are several ViT-s version in Refiners - for CLIP image encoder, DINOv2, etc) we decided to have multiple versions (no DRY) precisely because there are some subtleties here and there.

> But I can reformulate my perceiverResampler implementation to match the one for IP-adapter.

I think that would be great (at least, it's worth trying!). Thanks!

Reply:
got it, ill commit the new implementation

Reply:
```
            Latents(num_latents, width, device=device, dtype=dtype),
            fl.Residual(
                fl.UseContext("ella", "timestep_embedding"),
                fl.SiLU(),
                fl.Linear(time_embedding_dim, width, device=device, dtype=dtype),
            ),
```
after checking, I found that the difference between ip perceiverresampler and ella perceiverresampler is the use of the time embeddings, they first sum the Latents with a feedforward projection of the time_embedding. They also use the time_embedding component with LayerNorm in the AdaLayerNorm module. And finally they use a squared Relu.